### ISSUE TITLE
SLF001: access to private field of another instance of the same class

### ISSUE THREAD

**Issue by nktrnv** [astral-sh/ruff#9022]
Is this a bug or do I understand something wrong?

```python
class Foo:
    def __init__(self) -> None:
        self._value = 0
    
    def get_new_instance_with_another_value(self, value: int) -> "Foo":
        foo = Foo()
        foo._value = value  # SLF001 Private member accessed: `_value`
        return foo
```

For example, in Java, such code will be valid:

```java
public class Foo {
    private int value = 0;
    
    public Foo getNewInstanceWithAnotherValue(int value) {
        Foo foo = new Foo();
        foo.value = value; // No issues
        return foo;
    }
}
```

**Comment by tjkuson** [comment]
I think the rule checks for accessing an object's private names whilst outside the object, whereas Java checks for accessing a private name outside the class. Not sure if this is intended, I think you can make a reasonable argument for both behaviours…

**Comment by sbrudenell** [comment]
FWIW, it seems `SLF001` is already suppressed in some cases like `__lt__` (or, perhaps it's not processing correctly in the first place?)

```python3
class Wrap:
    def __init__(self, other: Any) -> None:
        if isinstance(other, Wrap):
            self._value = other._value  # SLF001 Private member accessed: `_value`
        else:
            self._value = other

    def __lt__(self, other: Wrap) -> None:
        return id(self._value) < id(other._value)
```
```
$ ruff check --select SLF001 /tmp/t.py
/tmp/t.py:4:27: SLF001 Private member accessed: `_value`
Found 1 error.
$ ruff --version
ruff 0.4.1
```

**Comment by echo-lalia** [comment]
Just going to add, but I get a similar false positive in this code:
``` Py
class Printer:
    _prntclrs = {
        'GRAY':'\033[90m',
    # ...
    }
    # ...
    @staticmethod
    def ctext(text:str, color:str) -> str:
        """Generate a colored string and return it."""
        color = Printer._prntclrs.get(color.upper(), "ENDC")
        return f"{color}{text}{Printer._prntclrs['ENDC']}"
```
Gives: `Private member accessed: `_prntclrs`Ruff[SLF001]`  
even though this is a static method of the Printer class, and this property only exists for use in this method.

**Comment by Satge96** [comment]
I would also love the change in the __new__ function. I'm fairly new to this, but if desired I could try to implement it.

**Comment by Tishka17** [comment]
I see that it is fixed for the code in the issue, but it not the only way to call "private" attrs 

```
class Foo:
    def __init__(self, other: "Foo | None") -> None:
        self._value = 0
        self._other = other

    def foo(self) -> None:
        if self._other is None:
            return
        self._value += self._other._value
```
